Regular expression models for literals

pkesseli opened the pull request:
As discussed in the workplace group I added support for source models for literals matching configurable regular expressions. The `models.md` documentation outlines what that would look like:

---
Example literal models:
```
[
  {
    "pattern": "SELECT \\*.*",
    "description": "Potential SQL Query",
    "sources": [
      {
        "kind": "SqlQuery"
      }
    ]
  },
  {
    "pattern": "AI[0-9A-Z]{16}",
    "description": "Suspected Google API Key",
    "sources": [
      {
        "kind": "GoogleAPIKey"
      }
    ]
  }
]
```

Example code:
```java
void testRegexSource() {
  String prefix = "SELECT * FROM USERS WHERE id = ";
  String aci = getAttackerControlledInput();
  String query = prefix + aci; // Sink
}

void testRegexSourceGoogleApiKey() {
  String secret = "AIABCD1234EFGH5678";
  sink(secret);
}
```
---

# Reviewer note
In order to test these changes, I added new classes and methods to `source/tests/integration/end-to-end/library_classes`, which causes a significant amount of unrelated changes to other tests' `expected_xxx.json`. I tried to constrain those changes to the commit `Update library_classes`.

yuhshin-oss commented on source/LiteralModel.h:
```diff
@@ -0,0 +1,81 @@
+  void join_with(const LiteralModel& other);
+
+  /**
+   * Indicats whether model contains any taints.
```
nit: spelling "Indicates"

pkesseli replied:
🐱‍👓

yuhshin-oss commented on source/LiteralModel.cpp:
```diff
@@ -0,0 +1,114 @@
+  return sources_.is_bottom();
+}
+
+static constexpr const char* PATTERN = "pattern";
+
+static constexpr const char* SOURCES = "sources";
```
nit: we generally don't define json keys as constants in our code base (reasoning being that it adds another level of indirection for something that is quite localized). We just use the string literal at the call-site directly

yuhshin-oss commented on source/LiteralModel.cpp:
```diff
@@ -0,0 +1,114 @@
+LiteralModel LiteralModel::from_json(
+    const Json::Value& value,
+    Context& context) {
+  JsonValidation::validate_object(value);
```
nit: I think check_unexpected_members already calls `validate_object`

yuhshin-oss commented on source/LiteralModel.cpp:
```diff
@@ -0,0 +1,114 @@
+static Taint create_taint(const std::string& pattern, const std::vector<TaintConfig>& sources) {
+  Taint all_sources;
+  for (const auto& source : sources) {
+    mt_assert(source.is_leaf());
```
If this assert fails, wouldn't it be a user error? (i.e. json with callee specified)

would be nicer to raise a JsonValidationError

pkesseli replied:
Yes, I think this would be a user error. I used `mt_assert` because that's what's used in `FieldModel` in a similar context:

```C++
void FieldModel::add_source(TaintConfig source) {
  mt_assert(source.is_leaf());
  check_taint_config_consistency(source, "source");
  add_source(Taint{std::move(source)});
}
```

I now added an additional check in `LiteralModel::from_json` where we have the JSON value for context:

```C++
if (!taintConfig.is_leaf())
  throw JsonValidationError{source, /* field */ "sources", "empty callee"};
```

So if we still violate the `mt_assert` in `create_taint`, it would be an application logic error.

yuhshin-oss commented on source/LiteralModel.cpp:
```diff
@@ -0,0 +1,114 @@
+  if (!sources_.is_bottom()) {
+    auto sources_value = Json::Value(Json::arrayValue);
+    for (const auto& source : sources_.frames_iterator()) {
+      mt_assert(!source.is_bottom());
```
similar to the above, I think this could be reported as a user error?

pkesseli replied:
This case is in `LiteralModel::to_json`, so there is no JSON user input being parsed here. This can only occur in the case of an application logic error, where we constructed a `LiteralModel` which cannot be serialised. After a quick browse I couldn't find any `::to_json` method throwing a `JsonValidationError`, I think they all use `mt_assert` in such cases.

However, we could e.g. throw an `std::invalid_argument` if preferred?

yuhshin-oss replied:
ah yeah, my bad, I think `mt_assert` is good here, thanks!

yuhshin-oss commented on source/Registry.cpp:
```diff
@@ -134,6 +148,16 @@ FieldModel Registry::get(const Field* field) const {
 
+LiteralModel Registry::get(const std::string_view literal) const {
+  LiteralModel result_model;
+  for (const auto& literal_model : literal_models_) {
+    if (literal_model.second.matches(literal)) {
```
nit: think we usually do:
```
for (const auto& [pattern, model] : literal_models_) {
  if (pattern.matches(literal)) {
    // ...
  }
}
```

yuhshin-oss commented on source/Registry.h:
```diff
@@ -87,6 +91,7 @@ class Registry final {
 
   mutable ConcurrentMap<const Method*, Model> models_;
   ConcurrentMap<const Field*, FieldModel> field_models_;
+  ConcurrentMap<std::string, LiteralModel> literal_models_;
```
looks like we're iterating over the models in the `get` call anyway so we probably don't gain much from a map... maybe use `ConcurrentSet` instead?

yuhshin-oss commented on source/tests/integration/end-to-end/code/literal_source/RegexSource.java:
```diff
@@ -0,0 +1,95 @@
+  // -------------------------------------------------------
+  //  Issues
+  // -------------------------------------------------------
+  public static void testRegexSource() {
```
nit: we usually have the word "issue" somewhere in the method name if we're expecting an issue there... so maybe `testRegexSourceIssue()` or something? (and `testRegexSourceNoMatchNoIssue()` when there are no issues expected)

yuhshin-oss reviewed (changes requested):
Thanks! Looks pretty good to me :D Just small comments/nits

pkesseli commented:
I also realised since I updated `library_classes` I also affected `source/tests/integration/end-to-end/code/check_cast_types`, which doesn't get run by default in `tests.yml`. I updated those expected JSON files in my latest push as well now.

I manually compiled `check_cast_types/KotlinCheckCast.kt` and added it to the JAR - not sure whether that's intended way of running this test, but it produced the same JSON output files again except for my new models in `library_classes`. 🙂

yuhshin-oss commented on source/ForwardTaintTransfer.cpp:
```diff
@@ -5,6 +5,8 @@
  * LICENSE file in the root directory of this source tree.
  */
 
+#include <regex>
```
Missed this before, but let's use <re2/re2.h> instead

yuhshin-oss replied:
Missed this before, but let's use <re2/re2.h> instead

yuhshin-oss commented on source/ForwardTaintTransfer.cpp:
```diff
@@ -1199,6 +1201,33 @@ void infer_output_taint(
+
+  const std::string_view literal{instruction->get_string()->str()};
+  const LiteralModel model{context->registry.get(literal)};
+  if (model.empty())
+    return false;
```
nit: style:

```
if (model.empty()) {
  return false;
}
```

yuhshin-oss commented on source/LiteralModel.h:
```diff
@@ -0,0 +1,81 @@
+
+#pragma once
+
+#include <regex>
```
Similarly, use <re2/re2.h> instead.

yuhshin-oss commented on source/LiteralModel.h:
```diff
@@ -0,0 +1,81 @@
+  Taint sources_;
+};
+
+}
```
nit:
```
} // namespace marianatrench
```

yuhshin-oss reviewed (changes requested):
Ugh, couple of clangtidy/format issues showing up internally that's too tedious to type all of them here. If you can address the RE2 one, we should be able to auto-format after importing

pkesseli commented:
> Ugh, couple of clangtidy/format issues showing up internally that's too tedious to type all of them here. If you can address the RE2 one, we should be able to auto-format after importing

Thanks, all addressed. I wasn't sure how to run the linter myself, otherwise I would have checked for that!

arthaud commented on source/ForwardTaintTransfer.cpp:
```diff
@@ -1199,6 +1199,34 @@ void infer_output_taint(
+  environment->write(
+      aliasing.result_memory_locations(),
+      model.sources(),
+      UpdateKind::Weak);
```
I think we should use a strong update here. This could matter if this is in a loop.

arthaud commented on source/LiteralModel.cpp:
```diff
@@ -0,0 +1,117 @@
+  for (const auto& source : sources) {
+    mt_assert(source.is_leaf());
+    check_taint_config_consistency(pattern, source);
+    all_sources.join_with(Taint{std::move(source)});
```
You are using `std::move` on a const ref, that does nothing. might as well remove that `std::move`.

pkesseli replied:
Thanks, I changed this to a const ref when I migrated to re2 and I missed the move.

arthaud commented on source/LiteralModel.cpp:
```diff
@@ -0,0 +1,117 @@
+  return all_sources;
+}
+
+LiteralModel::LiteralModel() : regex_(pattern_) { }
```
so we create a `LiteralModel` with an empty string and a pattern for the empty string?
I would prefer if this was more clear in the internal state, for instance we could use a `std::optional<re2::RE2>`.

pkesseli replied:
I now used an optional, and we do not match anything if default.

arthaud commented on source/LiteralModel.cpp:
```diff
@@ -0,0 +1,117 @@
+LiteralModel::LiteralModel() : regex_(pattern_) { }
+
+LiteralModel::LiteralModel(
+    std::string&& pattern,
```
We usually take by copy in those cases.
See https://www.youtube.com/watch?v=PNRju6_yn3o

Also, as a rule of thumb, let's be consistent with the rest of the codebase.

arthaud replied:
Well technically `std::string_view` is worse in the case where we move a `std::string`, it will make an extra copy.
but performance is not a problem here unless proven otherwise.

pkesseli replied:
I am not sure I am seeing an extra copy in that case? `std::string_view` doesn't have a constructor for `std::string &&`, so won't it just do the same as usual and just copy the pointer and length? And the constructor of `RE2` will always copy exactly once, irrespective of whether we use `const std::string &` or `const StringPiece &`/`std::string_view`, since `StringPiece` has a `string_view` constructor.

I thus defaulted to `std::string_view` since it doesn't require the caller to use `std::move`. Having said that, we invoke this constructor exactly once with an `rvalue` anyway, so I just changed it to an `std::string` to keep things simple and consistent. 🙂

arthaud commented on source/LiteralModel.cpp:
```diff
@@ -0,0 +1,117 @@
+}
+
+void LiteralModel::join_with(const LiteralModel& other) {
+  if (this != &other) {
+    sources_.join_with(other.sources_);
+  }
```
If the patterns are different I think it would make sense to set it to null/empty?

pkesseli replied:
Now checking if they have the same pattern, resetting `regex_` otherwise.

arthaud commented on source/LiteralModel.cpp:
```diff
@@ -0,0 +1,117 @@
+
+Json::Value LiteralModel::to_json(Context& context) const {
+  auto value = to_json(context.options->export_origins_mode());
+  value["position"] = context.positions->unknown()->to_json();
```
Not sure I understand why we want to export the empty position? It's just wasting space?

pkesseli replied:
As with a few other items, that's because I started from `FieldModel` and that's how it behaves:
https://github.com/facebook/mariana-trench/blob/f5c7c535f30ff554e8c5643988535f0736577d4e/source/FieldModel.cpp#L188

Happy to delete this line though. 🙂

arthaud replied:
Fine, let's keep it for consistency.

arthaud commented on source/LiteralModel.h:
```diff
@@ -0,0 +1,89 @@
+  Json::Value to_json(Context& context) const;
+
+ private:
+  std::string pattern_;
```
We don't need to store the pattern since `RE2` already does it. We could get it with `regex_.pattern()`.

pkesseli replied:
Thanks, this is still an artifact from using `std::regex` before I migrated to `RE2`.

arthaud commented on source/Registry.cpp:
```diff
@@ -176,13 +200,25 @@ void Registry::join_with(const FieldModel& field_model) {
 
+void Registry::join_with(const LiteralModel& literal_model) {
+  auto iterator = literal_models_.find(literal_model.pattern());
+  if (iterator != end(literal_models_)) {
```
nitpick: we usually use `literal_models_.end()`. Let's be consistent with the rest of the codebase.

arthaud commented on source/Registry.cpp:
```diff
@@ -298,14 +347,18 @@ void Registry::dump_models(
         auto writer = JsonValidation::compact_writer();
         for (std::size_t i = batch_size * batch; i < batch_size * (batch + 1) &&
-             i < models.size() + field_models.size();
+             i < models.size() + field_models.size() + literal_models.size();
              i++) {
           if (i < models.size()) {
             writer->write(models[i].to_json(context_), &batch_stream);
-          } else {
+          } else if (i < models.size() + field_models.size()) {
             writer->write(
                 field_models[i - models.size()].to_json(context_),
                 &batch_stream);
+          } else {
+            writer->write(
+                literal_models[i - models.size() - field_models.size()].to_json(context_),
+                &batch_stream);
```
pre-existing but we should probably rewrite this with a proper iterator. These index computations are error prone.

arthaud replied:
Just to be clear, I'm not expecting you to implement this. This could be done in another PR, or we could do this internally.

Anyway, I changed my mind here. Yes we could probably make this work with iterators but the fact that iterators hide the indexing will make this code unreadable.

What might make it easier to read is abstracting away "models + field_models + literal_models". We could have:
```
class ModelVector {
  public:
   std::size_t size() const { return _models.size() + _field_models.size() + _literal_models.size(); }

  Json::Value json_for_index(std::size_t i) {
    if (i < _models.size()) {
      return _models[i].to_json();
    }
    i -= _models.size();
    if (i < _field_models.size()) {
      return _field_models[i].to_json();
    }
    i -= _field_models.size();
    if (i < _literal_models.size()) {
      return _literal_models[i].to_json();
    }
    mt_unreachable("invalid index");
  }

  private:
    std::vector<Model> _models;
    std::vector<FieldModel> _field_models;
    std::vector<LiteralModel> _literal_models;
}
```

arthaud commented on source/Registry.h:
```diff
@@ -63,6 +65,7 @@ class Registry final {
   /* These are thread-safe. */
   Model get(const Method* method) const;
   FieldModel get(const Field* field) const;
+  LiteralModel get(std::string_view literal) const;
```
Could we change the name of this method?
When I read `get` I usually expect a `O(1)` but here we loop over all literal models.
Also, we might want another API latter that is an actual `get` for a given pattern.
I would call it `match_literal` or something.
Also nitpick: the comment above says "These are thread-safe" but your implementation of `get` is not thread-safe (iterating on a ConcurrentMap is not thread safe). I think it's fine though as long as there are no mutations in parallel.

pkesseli replied:
Technically not my comment, but definitely surprised to learn that redex' `ConcurrentMap` concurrent traversal isn't defined. Not what I'm used to from `concurrent_unordered_map`. 🙂 

Since I renamed the method, I now moved it away from those two `get` mehods and gave it a comment of its own.

arthaud commented on source/tests/integration/end-to-end/library_classes/StringBuilder.java:
```diff
@@ -0,0 +1,15 @@
+
+package java.lang;
+
+public class StringBuilder {
+
+  public StringBuilder append(String str) {
+    return this;
+  }
+}
```
Could all changes to library_classes be in a different PR? This causes a lot of changes in integration tests that are not related to the PR, making it hard to review.

arthaud replied:
Thanks, could you rebase this on top of master now that #140 was merged?

pkesseli replied:
Done!

arthaud reviewed (changes requested):
Sorry for coming up late on this, I saw a few things after @yuhshin-oss that I think are worth improving. Please bear with me.

arthaud commented on source/LiteralModel.cpp:
```diff
@@ -0,0 +1,120 @@
+}
+
+LiteralModel::LiteralModel(const LiteralModel& other)
+    : regex_(other.regex_ ? std::optional(other.regex_->pattern()) : std::nullopt), sources_(other.sources_) {
```
I guess this is because `re2::RE2` is not copyable? That's annoying.

pkesseli replied:
Correct:
```c++
  // Not copyable.
  // RE2 objects are expensive. You should probably use std::shared_ptr<RE2>
  // instead. If you really must copy, RE2(first.pattern(), first.options())
  // effectively does so: it produces a second object that mimics the first.
  RE2(const RE2&) = delete;
```

arthaud replied:
I think that's fine as is :)

arthaud commented on source/LiteralModel.cpp:
```diff
@@ -0,0 +1,120 @@
+    : regex_(other.regex_ ? std::optional(other.regex_->pattern()) : std::nullopt), sources_(other.sources_) {
+}
+
+std::string LiteralModel::pattern() const {
+  return regex_ ? regex_->pattern() : "(no pattern)";
+}
```
Could we return a `std::optional` instead? Not a fan of implicitly returning "no pattern".

arthaud commented on source/LiteralModel.cpp:
```diff
@@ -0,0 +1,120 @@
+}
+
+void LiteralModel::join_with(const LiteralModel& other) {
+  if (this != &other) {
```
nitpick: We generally prefer the early-return pattern, instead of nesting if blocks.
```suggestion
  if (this == &other) {
    return;
 }
```

arthaud commented on source/LiteralModel.cpp:
```diff
@@ -0,0 +1,120 @@
+Json::Value LiteralModel::to_json(const ExportOriginsMode export_origins_mode) const {
+  auto value = Json::Value(Json::objectValue);
+
+  value["pattern"] = pattern();
```
We should check beforehand if the pattern is empty

pkesseli replied:
Omitted now if the pattern is empty.

arthaud commented on source/Registry.cpp:
```diff
@@ -176,13 +200,25 @@ void Registry::join_with(const FieldModel& field_model) {
 }
 
+void Registry::join_with(const LiteralModel& literal_model) {
+  auto iterator = literal_models_.find(literal_model.pattern());
```
Maybe assert that the pattern is non-empty?

pkesseli replied:
Now using `mt_assert`:
```c++
const std::optional<std::string> pattern{literal_model.pattern()};
mt_assert(pattern);
```
